### PR TITLE
fix: not omitting result when empty

### LINKED ISSUE #102: Methods that return nothing must contain `result` field
We were trying to call a method that returns nothing but the library that we use  in the client rejects to deserialize the response.

After some investigation we found the following in [JSON RPC 2.0 spec](https://www.jsonrpc.org/specification#response_object):

> **result**
>   This member is REQUIRED on success.
>   This member MUST NOT exist if there was an error invoking the method.
>   The value of this member is determined by the method invoked on the Server.
>
> **error**
>   This member is REQUIRED on error.
>   This member MUST NOT exist if there was no error triggered during invocation.
>   The value for this member MUST be an Object as defined in section 5.1.

So on success the `result` field must exists.

The following unit tests are wrong:
https://github.com/filecoin-project/go-jsonrpc/blob/17a35772b24f59ceee25cc8d8177a2b255a659ac/rpc_test.go#L130-L133

### PR DESCRIPTION
**Pull request by distractedm1nd**
Closes #102 

A MarshalJSON method is required because the omitempty is conditional - if result could be set if error exists, the diff would only be changing the tag on the struct.

### CODE REVIEW

**Comment by distractedm1nd** [comment]
@Wondertan It does not have the same effect unfortunately - that was my first solution but it still violates the spec:

This member is REQUIRED on success.
This member MUST NOT exist if there was an error invoking the method.
The value of this member is determined by the method invoked on the Server.

It must exist on success (if error is not empty), even if the value is null. This is what I mean in the PR description

File: handler.go [hunk 1]
```diff
@@ -110,6 +110,26 @@ type response struct {
 	Error   *respError  `json:"error,omitempty"`
 }
 
+func (r response) MarshalJSON() ([]byte, error) {
+	// Result field MUST NOT exist if there was an error invoking the method.
+	// Result field is REQUIRED on success.
```
Comment (Wondertan) [inline handler.go:115]: Doc nit: add a line that this is from specs

**Review by Wondertan — APPROVED** [review]
Explanation makes sense. Thanks

**Review by magik6k — COMMENTED** [review]
Hey, sorry for taking so long to get to this.

Generally this direction looks good, but there's either a serious perf regression here or something is subtly wrong somewhere - this PR makes my tests hang, `go test -count=1 -v ./...` has now been running for a few mitutes with no signs of finishing, and they do pass in a few seconds on master.

**Comment by magik6k** [comment]
Fixed version of this was now merged in https://github.com/filecoin-project/go-jsonrpc/pull/107

**Pull request closed without merging**